Docs: Splits Kafka topic. Adds detailed example for kafka inputFormat (#11912)
apache/druid: docs/development/extensions-core/kafka-ingestion.md

id: kafka-ingestion
title: "Apache Kafka ingestion"

> sidebar_label: "Apache Kafka"

Reviewer: A small thing, but I'd suggest making the sidebar title to be the same as the page title here, just to make the distinction with the next two pages in the tree, which also start with "Apache Kafka" 

```suggestion
sidebar_label: "Apache Kafka ingestion"
```

  ~ under the License.
  -->

When you enable the Kafka indexing service, you can configure supervisors on the Overlord to manage the creation and lifetime of Kafka indexing tasks.

Kafka indexing tasks read events using Kafka's own partition and offset mechanism to guarantee exactly-once ingestion. The supervisor oversees the state of the indexing tasks to:
  - coordinate handoffs
  - manage failures
  - ensure that scalability and replication requirements are maintained.

This topic covers how to submit a supervisor spec to ingest data from Kafka. See the following for more information:
- For a reference of Kafka supervisor spec configuration options, see the [Kafka supervisor reference](./kafka-supervisor-reference.md).

> - For an operations reference information to help run and maintain Apache Kafka supervisors, see [Kafka supervisor operations](./kafka-supervisor-operations.md)

Reviewer: ```suggestion
- For operations reference information to help run and maintain Apache Kafka supervisors, see [Kafka supervisor operations](./kafka-supervisor-operations.md).
```

If your Kafka cluster enables consumer-group based ACLs, you can set `group.id` in `consumerProperties` to override the default auto generated group id.

## Load the Kafka indexing service

To use the Kafka indexing service, load the `druid-kafka-indexing-service` extension on both the Overlord and the MiddleManagers. See [Loading extensions](../extensions.md#loading-extensions) for instructions on how to configure extensions.

## Define a supervisor spec
Similar to the ingestion spec for batch ingestion, the supervisor spec configures the data ingestion for Kafka streaming ingestion. A supervisor spec has the following sections:
- `dataSchema` to specify the Druid datasource name, primary timestamp, dimensions, metrics, transforms, and any necessary filters.

> - `ioConfig` to configure Druid to connect to Kafka how to parse the data. Kafka-specific connection details go in the  `consumerProperties`. For more information, see the [Kafka supervisor reference](./kafka-supervisor-reference.md).

Reviewer: "..connect to Kafka _and_ how to parse the data"? Or to rewrite a bit more...

 ```suggestion
- `ioConfig` to configure Kafka connection settings and configure how Druid parses the data. Kafka-specific connection details go in the `consumerProperties`. For more information, see the [Kafka supervisor reference](./kafka-supervisor-reference.md).
```

> - `ioConfig` to configure Druid to connect to Kafka how to parse the data. Kafka-specific connection details go in the  `consumerProperties`. For more information, see the [Kafka supervisor reference](./kafka-supervisor-reference.md).

Reviewer 2: ```suggestion
- `ioConfig` to configure Druid to connect to Kafka how to parse the data. Kafka-specific connection details go in the `consumerProperties`. For more information, see the [Kafka supervisor reference](./kafka-supervisor-reference.md).
```

- `tuningConfig` to control various tuning parameters specific to each ingestion method.
For a full description of all the fields and parameters in a Kafka supervisor spec, see the [Kafka supervisor reference](./kafka-supervisor-reference.md).

> For information on how to configure the input format, see [Data formats](../../ingestion/data-formats.md).

Reviewer 3: This brings the "input format" concept up front with no explanation... maybe it is better to introduce it saying it is part of the `tuningConfig` and maybe succinctly summarize what it is for? (i.e. specify the data format of input data)

Author: Good catch! That belonged with `ioConfig`. Mentioned it there and moved the link.

The following sections contain examples to help you get started with supervisor specs.

### JSON input format supervisor spec example

> The following example demonstrates a supervisor spec for Kafka that uses the `JSON` input format. In this case Druid parses the message contents from the value field which are in JSON format:

Reviewer: The "which" is ambiguous, at first read, as to whether it refers to the value field or the message contents (more likely due to the verb tense). Maybe either "...the value field, which is in JSON format" .. or:

```suggestion
The following example demonstrates a supervisor spec for Kafka that uses the `JSON` input format. In this case Druid parses the message contents in JSON format from the value field:
```

}
```

### Kafka input format supervisor spec example
If you want to ingest data from other fields in addition to the Kafka message contents, you can use the `kafka` input format. The `kafka` input format lets you ingest:
- the message key field.
- message headers.
- the Kafka message timestamp.

> 

Reviewer 3: It also lets you ingest the kafka "value" (i.e. payload) of the kafka event. In Kafka it is more commonly to call each entry in the topic as "event" rather than "message". (https://kafka.apache.org/intro). In general a Kafka event can have: a key, a value, headers, and a timestamp. The Kafka Input Format of the supervisor spec lets you ingest any of one those. Including the event's value which is missing from the list you provide above.

Reviewer 3: Unfortunately I checked the previous terminology for "events" in previous Druid's Kafka documentation and Druid calls Kafka events "messages", so calling them "messages" here is ok for consistency with previous documentation.

> The Kafka inputFormat is currently designated as experimental.

For example, consider the following structure for a message that represents an fictitious wiki edit in a development environment:
- **Event headers**: {"environment": "development"}
- **Event key**: {"key: "wiki-edit"}
- **Event value**: \<JSON object with message contents of the change details\>
- **Event timestamp**: "Nov. 10, 2021 at 14:06"

When you use the `kafka` input format, you configure the way that Druid names the dimensions created from the Kafka message:

> - `headerLabelPrefix`: Supply a prefix to the Kafka headers to avoid any conflicts with named dimensions. The default is `kafka.header.`. Considering the header from the example, Druid maps the header to the following column: `kafka.header.enviornment`.

Reviewer 3: Spelling: `kafka.header.environment`

Reviewer: This one still needs to be done I think.

> - `timestampColumnName`:Supply a custom name for the Kafka timestamp in the Druid schema to avoid conflicts with other time columns. The default is `kafka.timestamp`.

Reviewer: ```suggestion
- `timestampColumnName`: Supply a custom name for the Kafka timestamp in the Druid schema to avoid conflicts with other time columns. The default is `kafka.timestamp`.
```

- `keyColumnName`: Supply the name for the Kafka key column in Druid. The default is `kafka.key`.
Additionally, you must provide information about how Druid should parse the data in the Kafka message:
- `headerFormat`: The default "string" decodes UTF8-encoded strings from the Kafka header. If you need another format, you can implement your own parser.

> - `keyFormat`: Takes a Druid `inputFormat` and uses the value for first key it finds. According to the example the value is "wiki-edit". It discards the key name in this case. If you store the key as a string, use the `CSV` input format. For example if you have simple string for the the key `wiki-edit`, you can use the following to parse the key:

Reviewer: ```suggestion
- `keyFormat`: Takes a Druid `inputFormat` and uses the value for the first key it finds. According to the example the value is "wiki-edit". It discards the key name in this case. If you store the key as a string, use the `CSV` input format. For example, if you have simple string for the the key `wiki-edit`, you can use the following to parse the key:
```

  ```
  "keyFormat": {
    "type": "csv",
    "hasHeaderRow": false,
    "findColumnsFromHeader": false,
    "columns": ["key"]
    } 
  ```
- `valueFormat`: Define how to parse the message contents. You can use any of the Druid input formats that work for Kafka.

For more information on data formats, see [Data formats](../../ingestion/data-formats.md).

> Finally, add the Kafka message columns to the `dimensionsSpec`. For the key and timestamp, you can use the dimension names you defined for `keyColumnName` and `timestampColumnName`. For header dimensions, supply the append the header key to the `headerLabelPrefix`. For example `kafka.header.environment`.

Reviewer 3: Reword this: "supply the append the header key"

      },
      "dimensionsSpec": {
        "dimensions": [
          {
            "type": "string",
            "name": "kafka.key"
          },
          {
            "type": "string",
            "name": "kafka.timestamp"
          },
          {
            "type": "string",
            "name": "kafka.header.environment"
          },
          "$schema",
          {
            "type": "long",
            "name": "id"
          },
          "type",
          {
            "type": "long",
            "name": "namespace"
          },
          "title",
          "comment",
          "user",]
        ]
      },
      "granularitySpec": {
        "queryGranularity": "none",
        "rollup": false,
        "segmentGranularity": "day"
      }
    }
  },
  "tuningConfig": {
    "type": "kafka"
  }
}
```
After Druid ingests the data, you can query the Kafka message columns as follows:
```unix
SELECT
  "kafka.header.environment",
  "kafka.key",
  "kafka.timestamp"
FROM "wikiticker"

kafka.header.environment  kafka.key	 kafka.timestamp
development               wiki-edit	 1636399229823
```
For more information, see [`kafka` data format](../../ingestion/data-formats.md#kafka).

> ## Submit a Supervisor Spec

Reviewer: ```suggestion
## Submit a supervisor spec
```

> Druid starts a supervisor for a dataSource when you submit a supervisor spec. You can use the Druid Load Data console or you can submit a supervisor spec to the following following endpoint:

Reviewer: ```suggestion
Druid starts a supervisor for a dataSource when you submit a supervisor spec. You can use the data loader in the Druid console or you can submit a supervisor spec to the following endpoint:
```
